### ISSUE TITLE
Using XSTATE features in user space applications

### ISSUE THREAD

**Issue by xuliker** [xuliker/kde#299]
Using XSTATE features in user space applications
在用户空间应用中使用XSTATE功能
================================================

The x86 architecture supports floating-point extensions which are
enumerated via CPUID. Applications consult CPUID and use XGETBV to
evaluate which features have been enabled by the kernel XCR0.

x86架构支持通过CPUID列举的浮点扩展。
应用程序查阅CPUID并使用XGETBV
来评估哪些特性已经被内核XCR0启用。 

Up to AVX-512 and PKRU states, these features are automatically enabled by
the kernel if available. Features like AMX TILE_DATA (XSTATE component 18)
are enabled by XCR0 as well, but the first use of related instruction is
trapped by the kernel because by default the required large XSTATE buffers
are not allocated automatically.

在AVX-512和PKRU状态下，如果有这些特性，内核会自动启用。
像AMX TILE_DATA（XSTATE组件18）这样的特性也被XCR0启用，

>> 但是相关指令的第一次使用会被内核困住，

因为在默认情况下，所需的大型XSTATE缓冲区不会被自动分配。 

Using dynamically enabled XSTATE features in user space applications
在用户空间应用程序中使用动态启用的XSTATE特性 
--------------------------------------------------------------------

The kernel provides an arch_prctl(2) based mechanism for applications to
request the usage of such features. The arch_prctl(2) options related to
this are:

内核提供了一个基于arch_prctl(2)的机制，
让应用程序请求使用这些特性。
与此相关的arch_prctl(2)选项有:

-ARCH_GET_XCOMP_SUPP

 arch_prctl(ARCH_GET_XCOMP_SUPP, &features);

 ARCH_GET_XCOMP_SUPP stores the supported features in userspace storage of
 type uint64_t. The second argument is a pointer to that storage.

-ARCH_GET_XCOMP_SUPP 
arch_prctl(ARCH_GET_XCOMP_SUPP, &features);
ARCH_GET_XCOMP_SUPP在用户空间存储uint64_t类型的支持特性。
第二个参数是一个指向该存储空间的指针。 

-ARCH_GET_XCOMP_PERM

 arch_prctl(ARCH_GET_XCOMP_PERM, &features);

 ARCH_GET_XCOMP_PERM stores the features for which the userspace process
 has permission in userspace storage of type uint64_t. The second argument
 is a pointer to that storage.

-ARCH_GET_XCOMP_PERM 
arch_prctl(ARCH_GET_XCOMP_PERM, &features); 
ARCH_GET_XCOMP_PERM将用户空间进程有权限的特性
存储在uint64_t类型的用户空间存储器中。第二个参数是一个指向该存储空间的指针。 

-ARCH_REQ_XCOMP_PERM

 arch_prctl(ARCH_REQ_XCOMP_PERM, feature_nr);

 ARCH_REQ_XCOMP_PERM allows to request permission for a dynamically enabled
 feature or a feature set. A feature set can be mapped to a facility, e.g.
 AMX, and can require one or more XSTATE components to be enabled.

 The feature argument is the number of the highest XSTATE component which
 is required for a facility to work.

- ARCH_REQ_XCOMP_PERM 
arch_prctl(ARCH_REQ_XCOMP_PERM, feature_nr); 
ARCH_REQ_XCOMP_PERM 允许为一个动态启用的特性或
一个特性集申请权限一个特征集可以映射到一个设施，例如AMX，
并且可以要求一个或多个XSTATE组件被启用。  

特征参数是设施工作所需的最高XSTATE组件的编号。 

When requesting permission for a feature, the kernel checks the
availability. The kernel ensures that sigaltstacks in the process's tasks
are large enough to accommodate the resulting large signal frame. It
enforces this both during ARCH_REQ_XCOMP_SUPP and during any subsequent
sigaltstack(2) calls. If an installed sigaltstack is smaller than the
resulting sigframe size, ARCH_REQ_XCOMP_SUPP results in -ENOSUPP. Also,
sigaltstack(2) results in -ENOMEM if the requested altstack is too small
for the permitted features.

当请求一个特性的许可时，内核会检查其可用性。
内核确保进程任务中的sigaltstacks足够大，以容纳所产生的大信号帧。
它在ARCH_REQ_XCOMP_SUPP期间和任何后续的sigaltstack(2)调用期间都强制执行。
如果安装的sigaltstack小于产生的sigframe大小，
ARCH_REQ_XCOMP_SUPP的结果就是-ENOSUPP。
另外，如果请求的altstack对于允许的特性来说太小，sigaltstack(2)会导致-ENOMEM。 

Permission, when granted, is valid per process. Permissions are inherited
on fork(2) and cleared on exec(3).

权限，当被授予时，对每个进程有效。
权限在fork(2)中被继承，在exec(3)中被清除。 

The first use of an instruction related to a dynamically enabled feature is
trapped by the kernel. The trap handler checks whether the process has
permission to use the feature. If the process has no permission then the
kernel sends SIGILL to the application. If the process has permission then
the handler allocates a larger xstate buffer for the task so the large
state can be context switched. In the unlikely cases that the allocation
fails, the kernel sends SIGSEGV.

第一次使用与动态启用的特性相关的指令会被内核捕获。
陷阱处理程序检查该进程是否有使用该特性的权限。
如果进程没有权限，那么内核会向应用程序发送SIGILL。
如果进程有权限，那么处理程序将为任务分配一个更大的xstate缓冲区，
这样大的状态就可以被上下文切换。
在不太可能的情况下，如果分配失败，内核会发送SIGSEGV。 

Dynamic features in signal frames
信号帧中的动态特性 
---------------------------------

Dynamcally enabled features are not written to the signal frame upon signal
entry if the feature is in its initial configuration.  This differs from
non-dynamic features which are always written regardless of their
configuration.  Signal handlers can examine the XSAVE buffer's XSTATE_BV
field to determine if a features was written.

如果动态功能处于初始配置状态，那么在信号进入时，动态功能不会被写入信号帧。 
这与非动态特性不同，后者无论其配置如何都会被写入。 
信号处理程序可以检查XSAVE缓冲区的XSTATE_BV字段，以确定一个特征是否被写入。

**Comment by xuliker** [comment]
x86/cpufeatures: Add eXtended Feature Disabling (XFD) feature bit

Intel's eXtended Feature Disable (XFD) feature is an extension of the XSAVE
architecture. XFD allows the kernel to enable a feature state in XCR0 and
to receive a #NM trap when a task uses instructions accessing that state.

英特尔的eXtended Feature Disable（XFD）功能是XSAVE架构的一个扩展。
XFD允许内核在XCR0中启用一个特性状态，
--> 并在任务使用访问该状态的指令时收到一个#NM陷阱。

This is going to be used to postpone the allocation of a larger XSTATE
buffer for a task to the point where it is actually using a related
instruction after the permission to use that facility has been granted.

这要用于推迟为任务分配更大的XSTATE缓冲区，
直到任务在获得使用该设施的许可后实际使用相关指令。

XFD is not used by the kernel, but only applied to userspace. This is a
matter of policy as the kernel knows how a fpstate is reallocated and the
XFD state.

--> XFD不被内核使用，而只应用于用户空间。
这是一个政策问题，因为内核知道fpstate是如何重新分配的，以及XFD的状态。 

The compacted XSAVE format is adjustable for dynamic features. Make XFD
depend on XSAVES.

压缩后的XSAVE格式可以调整为动态特性。使XFD依赖于XSAVES。

**Comment by xuliker** [comment]
x86/fpu: Update XFD state where required
x86/fpu: 在需要时更新XFD状态  

--> The IA32_XFD_MSR allows to arm #NM traps for XSTATE components which are enabled in XCR0. 
The register has to be restored before the tasks XSTATE is restored. 
The life time rules are the same as for FPU state.

IA32_XFD_MSR允许为XSTATE组件设置#NM陷阱，这些组件在XCR0中被启用。
在任务的XSTATE被恢复之前，该寄存器必须被恢复。恢复。寿命规则与FPU的状态相同。

XFD is updated on return to userspace only when the FPU state of the task
is not up to date in the registers. It's updated before the XRSTORS so
that eventually enabled dynamic features are restored as well and not
brought into init state.

只有当任务的FPU状态在寄存器中不是最新时，
XFD才会在返回用户空间时被更新。在寄存器中不是最新的。
它在XRSTORS之前被更新，所以最终启用的动态特性也会被恢复，
而不是 被带入初始状态。 

Also in signal handling for restoring FPU state from user space the
correctness of the XFD state has to be ensured.

此外，在从用户空间恢复FPU状态的信号处理中，
XFD状态的正确性也是如此。XFD状态的正确性必须得到保证。 

Add it to CPU initialization and resume as well.
在CPU的初始化和恢复中也要加上这一点。

**Comment by xuliker** [comment]
Dynamically enabled features are enabled in XCR0, but
usage requires also that the corresponding bits in XFD
are cleared.  If the bits are set then using a related
instruction will raise #NM. This allows to do the
allocation of the larger FPU buffer lazy from #NM or if
the task has no permission to kill it which would happen
via #UD if the feature is disabled in XCR0.

动态启用的功能在XCR0中被启用，但使用时还需要清除XFD中的相应位。 
如果这些位被设置，那么使用相关的指令将引发#NM。
这允许从#NM中分配较大的FPU缓冲区，或者如果任务没有权限杀死它，
这将通过#UD发生，如果该功能在XCR0中被禁用。 

XFD state is following the same life time rules as
XSTATE and to restore state correctly XFD has to be
updated before XRSTORS otherwise the component would
stay in or go into init state even if the bits are set
in fpstate::regs::xsave::xfeatures.

XFD状态遵循与XSTATE相同的寿命规则，为了正确恢复状态，
XFD必须在XRSTORS之前更新，否则组件将保持在或进入初始状态，
即使在FPstate::regs::xsave::xfeatures中设置了位。

**Comment by xuliker** [comment]
x86/fpu/xstate: Add fpstate_realloc()/free()
x86/fpu/xstate。增加 fpstate_realloc()/free()

The fpstate embedded in struct fpu is the default state for storing the FPU
registers. It's sized so that the default supported features can be stored.
For dynamically enabled features the register buffer is too small.

嵌入到结构 fpu 中的 fpstate 是用于存储 FPU 寄存器的默认状态。
它的大小使得默认支持的功能可以被存储。
对于动态启用的功能，寄存器缓冲区太小了。 

The #NM handler detects first use of a feature which is disabled in the
XFD MSR. After handling permission checks it recalculates the size for
kernel space and user space state and invokes fpstate_realloc() which
tries to reallocate fpstate and install it.

#NM处理程序检测首次使用在XFD MSR中被禁用的特性。在处理权限检查后，
它重新计算内核空间和用户空间状态的大小，并调用fpstate_realloc()，
试图重新分配fpstate并安装它。 

Provide the allocator function which checks whether the current buffer size
is sufficient and if not allocates one. If allocation is successful the new
fpstate is initialized with the new features and sizes and the now enabled
features is removed from the task's XFD mask.

提供分配器函数，它检查当前的缓冲区大小是否足够，如果不够就分配一个。
如果分配成功，新的fpstate被初始化为新的功能和大小，
现在启用的功能被从任务的XFD掩码中移除。

realloc_fpstate() uses vzalloc(). If use of this mechanism grows to
re-allocate buffers larger than 64KB, a more sophisticated allocation
scheme that includes purpose-built reclaim capability might be justified.

realloc_fpstate()使用vzalloc()。
如果这个机制的使用增长到重新分配大于64KB的缓冲区，
一个更复杂的分配方案可能是合理的，包括专门的回收能力。

**Comment by xuliker** [comment]
x86/fpu/amx: Define AMX state components and have it used for boot-time checks
x86/fpu/amx: 定义AMX状态组件并让它用于启动时检查  

The XSTATE initialization uses check_xstate_against_struct() to sanity
check the size of XSTATE-enabled features. AMX is a XSAVE-enabled feature,
and its size is not hard-coded but discoverable at run-time via CPUID.

XSTATE初始化使用check_xstate_against_struct()来理智地检查XSTATE启用的特性的大小。
AMX是一个XSAVE启用的特性，它的大小不是硬编码的，而是在运行时可以通过CPUID发现。 

The AMX state is composed of state components 17 and 18, which are all user
state components. The first component is the XTILECFG state of a 64-byte
tile-related control register. The state component 18, called XTILEDATA,
contains the actual tile data, and the state size varies on
implementations. The architectural maximum, as defined in the CPUID(0x1d,
1): EAX[15:0], is a byte less than 64KB. The first implementation supports
8KB.

AMX状态由状态组件17和18组成，它们都是用户状态组件。
第一个组件是一个64字节的瓦片相关控制寄存器的XTILECFG状态。
状态组件18，称为XTILEDATA，包含实际的瓦片数据，
状态大小在实现上有所不同。
架构上的最大值，在CPUID(0x1d, 1)中定义。EAX[15:0]，
是一个小于64KB的字节。第一个实现支持8KB。 

Check the XTILEDATA state size dynamically. The feature introduces the new
tile register, TMM. Define one register struct only and read the number of
registers from CPUID. Cross-check the overall size with CPUID again.

动态地检查XTILEDATA状态的大小。
该功能引入了新的瓦片寄存器TMM。
只定义一个寄存器结构，从CPUID读取寄存器的数量。
再次与CPUID交叉检查整体大小。

**Comment by xuliker** [comment]
x86/fpu: Calculate the default sizes independently

When dynamically enabled states are supported the maximum and default sizes
for the kernel buffers and user space interfaces are not longer identical.

当支持动态启用状态时，内核缓冲区和用户空间接口的最大和默认大小不再是相同的

Put the necessary calculations in place which only take the default enabled
features into account

把必要的计算放在只考虑到默认启用的功能的地方。

**Comment by xuliker** [comment]
x86/fpu: Add XFD handling for dynamic states
x86/fpu: 增加XFD对动态状态的处理  

To handle the dynamic sizing of buffers on first use the XFD MSR has to be
armed. Store the delta between the maximum available and the default
feature bits in init_fpstate where it can be retrieved for task creation.

为了处理首次使用时缓冲区的动态大小，XFD MSR必须被激活。
在init_fpstate中存储最大可用功能位和默认功能位之间的差值，
在任务创建时可以检索到它。 

If the delta is non zero then dynamic features are enabled. This needs also
to enable the static key which guards the XFD updates. This is delayed to
an initcall because the FPU setup runs before jump labels are initialized.

如果delta不为零，那么动态特性就被启用。这也需要启用静态密钥，
以保护XFD的更新。这被延迟到initcall，
因为FPU设置在跳转标签被初始化之前运行。

**Comment by xuliker** [comment]
selftests/x86/amx: Add test cases for AMX state management
selftests/x86/amx: 增加AMX状态管理的测试案例  

AMX TILEDATA is a very large XSAVE feature.  It could have caused
nasty XSAVE buffer space waste in two places:
AMX TILEDATA是一个非常大的XSAVE功能。 
它可能在两个地方造成讨厌的XSAVE缓冲区空间浪费。 

===> * Signal stacks
===> * Kernel task_struct->fpu buffers

* 信号堆栈 
* 内核task_struct->fpu缓冲区  

To avoid this waste, neither of these buffers have AMX state by
default.  The non-default features are called "dynamic" features.

为了避免这种浪费，这两个缓冲区默认都没有AMX状态。
这些非默认的特性被称为 "动态 "特性。 

There is an arch_prctl(ARCH_REQ_XCOMP_PERM) which allows a task
to declare that it wants to use AMX or other "dynamic" XSAVE
features.  This arch_prctl() ensures that sufficient sigaltstack
space is available before it will succeed.  It also expands the
task_struct buffer.

有一个arch_prctl(ARCH_REQ_XCOMP_PERM)允许任务
声明它想使用AMX或其他 "动态 "XSAVE特性。 
这个arch_prctl()确保有足够的sigaltstack空间才能成功。 它也扩展了task_struct缓冲区。 

Functions of this test:
这个测试的功能。

* Test arch_prctl(ARCH_REQ_XCOMP_PERM).  Ensure that it checks for
proper sigaltstack sizing and that the sizing is enforced for
future sigaltstack calls.
* Ensure that ARCH_REQ_XCOMP_PERM is inherited across fork()
* Ensure that TILEDATA use before the prctl() is fatal
* Ensure that TILEDATA is cleared across fork()

* 测试arch_prctl(ARCH_REQ_XCOMP_PERM)。 
确保它检查正确的sigaltstack大小，并且在未来的sigaltstack调用中强制执行该大小。
* 确保ARCH_REQ_XCOMP_PERM被继承到fork()中。
* 确保在prctl()之前使用TILEDATA是致命的。
* 确保TILEDATA在fork()中被清空。 

Note: Generally, compiler support is needed to do something with
AMX.  Instead, directly load AMX state from userspace with a
plain XSAVE.  Do not depend on the compiler.

注意：一般来说，需要编译器支持才能对AMX做一些事情。 
相反，直接用普通的XSAVE从用户空间加载AMX状态。 
不要依赖编译器。

**Comment by xuliker** [comment]
selftests/x86/amx: Add context switch test
selftests/x86/amx: 增加上下文切换测试  

XSAVE state is thread-local.  The kernel switches between thread
state at context switch time.  Generally, running a selftest for
a while will naturally expose it to some context switching and
and will test the XSAVE code.

XSAVE状态是线程本地的。 内核在上下文切换时在线程状态之间切换。 
一般来说，运行一段时间的selftest自然会暴露在一些上下文切换中，
并且会测试XSAVE代码。

Instead of just hoping that the tests get context-switched at
random times, force context-switches on purpose.  Spawn off a few
userspace threads and force context-switches between them.
Ensure that the kernel correctly context switches each thread's
unique AMX state.

与其希望测试在随机的时间内被切换到上下文，
不如故意强制进行上下文切换。 
生成几个用户空间线程，在它们之间强制进行上下文切换。
确保内核正确切换每个线程的AMX状态。

**Comment by xuliker** [comment]
x86/fpu: Optimize out sigframe xfeatures when in init state
x86/fpu: 在初始状态下，优化出sigframe的xfeatures。 

tl;dr: AMX state is ~8k.  Signal frames can have space for this
~8k and each signal entry writes out all 8k even if it is zeros.
Skip writing zeros for AMX to speed up signal delivery by about
4% overall when AMX is in its init state.

AMX状态是~8k。 信号帧可以有这个~8k的空间，
每个信号条目都会写出所有的8k，即使是零。
当AMX处于初始状态时，跳过为AMX写零来加快信号传递的整体速度，大约4%。 

This is a user-visible change to the sigframe ABI.
这是对sigframe ABI的一个用户可见的改变。

== Hardware XSAVE Background ==

XSAVE state components may be tracked by the processor as being
in their initial configuration.  Software can detect which
features are in this configuration by looking at the XSTATE_BV
field in an XSAVE buffer or with the XGETBV(1) instruction.

XSAVE状态的组件可以被处理器追踪到它们的初始配置中。 
软件可以通过查看XSAVE缓冲区的XSTATE_BV字段或
使用XGETBV(1)指令来检测哪些功能处于这种配置。 

Both the XSAVE and XSAVEOPT instructions enumerate features s
being in the initial configuration via the XSTATE_BV field in the
XSAVE header,  However, XSAVEOPT declines to actually write
features in their initial configuration to the buffer.  XSAVE
writes the feature unconditionally, regardless of whether it is
in the initial configuration or not.

==> XSAVE和XSAVEOPT指令通过XSAVE头中的
==> XSTATE_BV字段来列举处于初始配置中的特征，
==> 然而，XSAVEOPT拒绝将初始配置中的特征实际写入缓冲区。 
==> XSAVE无条件地写入特征，不管它是否在初始配置中。 

Basically, XSAVE users never need to inspect XSTATE_BV to
determine if the feature has been written to the buffer.
XSAVEOPT users *do* need to inspect XSTATE_BV.  They might also
need to clear out the buffer if they want to make an isolated
change to the state, like modifying one register.

基本上，XSAVE用户不需要检查XSTATE_BV来确定特征是否已经被写入缓冲区。
XSAVEOPT用户*需要检查XSTATE_BV。 如果他们想对状态做一个孤立的改变，
比如修改一个寄存器，他们可能还需要清除缓冲区。

== Software Signal / XSAVE Background ==
 ==软件信号/XSAVE背景 ==  

Signal frames have historically been written with XSAVE itself.
Each state is written in its entirety, regardless of being in its
initial configuration.

历史上，信号帧是用XSAVE本身来写的。
每个状态都被完整地写入，不管是否处于初始配置。 

In other words, the signal frame ABI uses the XSAVE behavior, not
the XSAVEOPT behavior.
换句话说，信号帧ABI使用XSAVE行为，而不是XSAVEOPT行为。 

== Problem ==

This means that any application which has acquired permission to
use AMX via ARCH_REQ_XCOMP_PERM will write 8k of state to the
signal frame.  This 8k write will occur even when AMX was in its
initial configuration and software *knows* this because of
XSTATE_BV.

这意味着任何通过ARCH_REQ_XCOMP_PERM获得使用AMX权限的
应用程序将向信号帧写入8k的状态。 
这个8k的写入将发生在AMX处于初始配置的时候，
而且由于XSTATE_BV的存在，软件*知道*这个。 

This problem also exists to a lesser degree with AVX-512 and its
2k of state.  However, AVX-512 use does not require
ARCH_REQ_XCOMP_PERM and is more likely to have existing users
which would be impacted by any change in behavior.

这个问题也存在于AVX-512和它的2k状态中，程度较轻。 
然而，AVX-512的使用不需要ARCH_REQ_XCOMP_PERM，
而且更有可能有现有的用户，他们会受到任何行为变化的影响。

== Solution ==

Stop writing out AMX xfeatures which are in their initial state
to the signal frame.  This effectively makes the signal frame
XSAVE buffer look as if it were written with a combination of
XSAVEOPT and XSAVE behavior.  Userspace which handles XSAVEOPT-
style buffers should be able to handle this naturally.

==> 停止写出处于初始状态的AMX xfeatures到信号帧。 
==> 这有效地使信号帧XSAVE缓冲区看起来像是用XSAVEOPT和XSAVE行为的组合写的。 
==> 处理XSAVEOPT-风格缓冲区的用户空间应该能够自然地处理这个问题。 

For now, include only the AMX xfeatures: XTILE and XTILEDATA in
this new behavior.  These require new ABI to use anyway, which
makes their users very unlikely to be broken.  This XSAVEOPT-like
behavior should be expected for all future dynamic xfeatures.  It
may also be extended to legacy features like AVX-512 in the
future.

现在，只包括AMX的xfeatures。XTILE和XTILEDATA在这个新行为中。 
这些都需要新的ABI来使用，这使得他们的用户很不可能被破坏。 
这种类似XSAVEOPT的行为应该被期待用于所有未来的动态xfeatures。

Only attempt this optimization on systems with dynamic features.
Disable dynamic feature support (XFD) if XGETBV1 is unavailable
by adding a CPUID dependency.

==> 它也可能在未来被扩展到像AVX-512这样的传统特性。 
==> 只在有动态特性的系统上尝试这种优化。如果XGETBV1不可用，
==> 则通过添加CPUID依赖来禁用动态特性支持（XFD）。

This has been measured to reduce the *overall* cycle cost of
signal delivery by about 4%.

据测量，这可以将信号传递的*总体*周期成本降低约4%。

**Comment by xuliker** [comment]
**Intel Architecture Day 2021 Demo- Advanced Matrix Extensions (AMX).mp4**

amx is a hardware block in the sapphire rapid cpu with the new expandable two-dimensional register file and a new matrix multiply instructions to enhance performance for a variety of deep learning workloads for both inference and training as you mentioned we can do more matrix multiplies per clock cycle so we can process data faster here in our lab we have a sapphire rapid server running an internally optimized general matrix multiply gem kernel on the left hand side we are running without amx and on the right hand side we are running with the amx extensions with amx's ability to do more matrix multiplies per clock cycle you can see that we are executing the gem kernel approximately 7.8 x faster with advanced matrix extensions while this demo highlights a highly efficient gem kernel to show the architectural capabilities of this platform we expect substantial performance gain across the ai work list for both training and inference 

amx是SPR CPU中的一个硬件模块，具有新的可扩展的二维寄存器文件和新的矩阵乘法指令，以提高推理和训练等各种深度学习工作负载的性能。 

正如你所提到的，我们可以在每个时钟周期内进行更多的矩阵乘法，因此我们可以更快地处理数据。 

在我们的实验室里，我们有一个SPR服务器，运行内部优化的通用矩阵乘法GEM内核，在左边，我们没有运行AMX，在右边，我们正在运行AMX扩展，AMX能够在每个时钟周期内进行更多的矩阵乘法，你可以看到，我们执行GEM内核的速度大约是7. 8倍的速度。 

这个演示突出了高效的gem内核，展示了这个平台的架构能力，我们期望在整个ai工作列表中，训练和推理的性能都能得到大幅提升。
